### PR TITLE
Default to 21.05, standard shell, binary cache aggregation, elimination of etwa alte scheiße 😌

### PR DESCRIPTION
**Pull request by kreisys**
So you can put a `nix.conf` fragment in `NIX_CONFIG` environment variable so that way there's no need to meddle with the user's config files. En plus, they added the `extra-` prefix that allows appending to whatever's been set previously. You can keep appending to `NIX_CONFIG`, repeating, for example, `extra-substituters = <whatever>` and it would aggregate. See https://github.com/input-output-hk/midnight-ops/blob/nixconf/flake.nix#L37-L46 and https://github.com/input-output-hk/bitte/blob/genshell/pkgs/bitte-shell.nix#L36-L41

### CODE REVIEW

File: flake.nix [hunk 1]
```diff
@@ -30,18 +27,18 @@
...
-  outputs = { self, hydra-provisioner, nixpkgs, utils, bitte-cli, nixpkgs-2105, ... }@inputs:
+  outputs = { self, hydra, hydra-provisioner, nixpkgs, utils, bitte-cli, ... }@inputs:
   let
     lib = import ./lib { inherit (nixpkgs) lib; };
   in utils.lib.simpleFlake rec {
     inherit lib nixpkgs;
 
     systems = [ "x86_64-linux" ];
 
+    preOverlays = [ bitte-cli hydra-provisioner hydra ];
```
Comment (nrdxp) [inline flake.nix:38]: does this need to be exported?

File: overlay.nix [hunk 2]
```diff
@@ -64,6 +59,8 @@ in final: prev: {
   haproxy-cors = prev.callPackage ./pkgs/haproxy-cors.nix { };
 
   devShell = final.callPackage ./pkgs/dev-shell.nix { };
+  bitteShell = final.callPackage ./pkgs/bitte-shell.nix { };
+  bitteSchoell = final.bitteShell;
```
Comment (nrdxp) [inline overlay.nix:63]: What's the point of the redundant shell package?

**Review by nrdxp — COMMENTED** [review]
Overall looks like a lot of good cleanups and fixes. Just had a few questions, but other than that, it looks good.

File: overlay.nix (hunk 2 above)
Comment (kreisys) [reply overlay.nix:63]: If you mean shell/schoell then that was just for my own lols. Bitte schön, shell, schöll. Meant to remove it but got distracted by popsicles.
If you mean devShell/bitteShell then devShell is this repo's devShell and bitteShell is a devShell template for consumer flakes (`*-ops`es).

File: lib/mk-hashi-stack.nix [hunk 3]
```diff
@@ -1,9 +1,12 @@
...
+{ flake, domain # E.g. "mantis.ws"
 , dockerRegistry ? "docker." + domain, dockerRole ? "developer"
 , vaultDockerPasswordKey ? "kv/nomad-cluster/docker-developer-password" }:
 
 let
-  lib = pkgs.lib;
+  inherit (flake.inputs) nixpkgs;
+  inherit (nixpkgs) lib;
+
+  pkgs = nixpkgs.legacyPackages.x86_64-linux.extend flake.overlay;
```
Comment (jonringer) [inline lib/mk-hashi-stack.nix:9]: the `rootDir` -> `flake` and overlay change are breaking changes. Not to mention that only applying one overly is probably not going to be sufficient for some other repos. 4 are used for one of our repositories.

File: flake.nix (hunk 1 above)
Comment (kreisys) [reply flake.nix:38]: https://github.com/kreisys/flake-utils/blob/master/simpleFlake.nix#L16-L19

`preOverlays` is just a compatibility alias for `overlays`. The singular `overlay` is just a convenience and it gets appended to `overlays` (which, in turn, default to `[]`). In the end, the `overlay` that's exposed in the attrset that is returned by the `simpleFlake` function (which happens to have an argument of the same name), is a single one, composed of all the overlays that had been passed into `simpleFlake` using the `composeManyExtensions` library function from `nixpkgs.lib`. Y'all need to give me more credit 🙃

File: lib/mk-hashi-stack.nix (hunk 3 above)
Comment (kreisys) [reply lib/mk-hashi-stack.nix:9]: https://github.com/kreisys/flake-utils/blob/master/simpleFlake.nix#L70

File: flake.nix (hunk 1 above)
Comment (nrdxp) [reply flake.nix:38]: thanks for heads up :+1:

File: pkgs/bitte-shell.nix [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    sops
+    vault
+    openssl
+    cfssl
+    nixfmt
+    awscli
+    nomad
+    consul
+    consul-template
+    python38Packages.pyhcl
```
Comment (manveru) [inline pkgs/bitte-shell.nix:64]: we don't use this.

**Review by manveru — CHANGES_REQUESTED** [review]
Overall not bad :)

File: modules/nomad-autoscaler.nix [hunk 5]
```diff
@@ -600,6 +600,11 @@ in {
       restartIfChanged = false;
       reloadIfChanged = true;
 
+      unitConfig = {
+        StartLimitInterval = "20s";
+        StartLimitBurst = 10;
+      };
```
[the remark this answers is not in the document]
Comment (kreisys) [reply modules/nomad-autoscaler.nix:606]: Oh didn't know that existed. I changed it in response to a deprecation warning iirc

File: pkgs/bitte-shell.nix [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+in mkShell ({
+  # for bitte-cli
+  LOG_LEVEL = "debug";
+
+  NIX_CONFIG = ''
+        extra-experimental-features = nix-command flakes ca-references recursive-nix
+        extra-substituters = https://hydra.iohk.io
+        extra-trusted-public-keys = hydra.iohk.io:f/Ea+s+dFdN+3Y/G+FDgSq+a5NEWhJGzdjvKNGv0/EQ=
+        bash-prompt = [bitte]
+  '' + (lib.optionalString (nixConfig != null) nixConfig);
```
[the remark this answers is not in the document]
Comment (nrdxp) [reply pkgs/bitte-shell.nix:41]: is the `nixConfig` flake attribute fully working yet? If so, we could dump these there, to set _before_ evaluation of the flake.

File: pkgs/bitte-shell.nix (hunk 4 above)
Comment (kreisys) [reply pkgs/bitte-shell.nix:64]: This was mostly copypasted from the old darwin branch so there must be some outdated stuff.
So remove `nomad`, `consul`, `consul-template` and whatever that python thing is?

File: pkgs/bitte-shell.nix (hunk 6 above)
[the remark this answers is not in the document]
Comment (manveru) [reply pkgs/bitte-shell.nix:41]: So i think .envrc is the better place to set this.

File: pkgs/bitte-shell.nix (hunk 4 above)
Comment (manveru) [reply pkgs/bitte-shell.nix:64]: We can keep the others, just pyhcl won't be used.

File: pkgs/bitte-shell.nix (hunk 6 above)
[the remark this answers is not in the document]
Comment (nrdxp) [reply pkgs/bitte-shell.nix:41]: Yes, according to [this comment](https://github.com/nix-community/rnix-lsp/pull/38), it wouldn't work for nix daemon users anyway. I just grepped master and it's still a FIXME

[the remark this answers is not in the document]
Comment (kreisys) [reply pkgs/bitte-shell.nix:41]: Only issue with `.envrc` is that it's another thing the enduser has to install. Another option I was thinking about is to keep the `devShell` minimal, consisting of only packages that are available in the mainline nixpkgs, and alias `bitte` to `nix run .#bitte`

File: pkgs/bitte-shell.nix (hunk 4 above)
Comment (nrdxp) [reply pkgs/bitte-shell.nix:64]: I was gonna say that I use the nomad cli quite a bit :laughing:

File: pkgs/bitte-shell.nix (hunk 6 above)
[the remark this answers is not in the document]
Comment (manveru) [reply pkgs/bitte-shell.nix:41]: None of the devs would use `bitte` anyway, it's only `iogo` that's a problem (though of course that's in the same package, so my point may be moot).
I think it's fine to require direnv and just pass all required options to the invocation of `nix print-dev-env` in there.
Restricting the devshell to the ones in upstream nixpkgs is likely to cause headaches down the line, so I'm not in favor of that.

**Pull request merged**